### ISSUE TITLE
Slow Git Pre-Receive Hook Performance

### ISSUE THREAD

**Issue by AJH16** [jakubgarfield/Bonobo-Git-Server#808]
I am having a performance related issue with using Git hooks.  If I run the server with no hook, all is good.  If I run the hook locally and run against the file system for the push, the hook runs successfully and completes very quickly.  If I run on Bonobo, the CopyTo on line 97 of GitServiceExecutor.cs takes an exceptionally long time to complete on a call to git.exe with arguments "receive-pack --stateless-rpc "C:\GitRepoServer\Bonobo.Git.Server\App_Data\Repositories\HookTest".  After about 30 seconds or so, the command finishes and it continues normally.   For this test case, the hook is simply calling an executable that writes 10 lines to stdout and 10 lines to stderr and returning -1;  Presumably something isn't closing right, but I can't tell what.

**Comment by willdean** [comment]
Does it make a difference if you don't write to stderr?

**Comment by AJH16** [comment]
It does not seem to from previous hooks that were doing more, but I'll verify tomorrow.On Jun 6, 2018 5:19 PM, Will Dean <notifications@github.com> wrote:Does it make a difference if you don't write to stderr?

**Comment by AJH16** [comment]
Ok, actually, I just realized that the success/failure issue may be unrelated.  Looks like the file system push is now succeeding and not getting any log messages, so something odd is going on there that I need to troubleshoot.  Not sure what's going on with the Bonobo version going so slow though.

**Comment by AJH16** [comment]
If you saw the previous message in e-mail, it was a dumb error on my part.  I hadn't renamed the file away from being a txt file after renaming it to post here.  So file system version is working again correctly, showing the return and responding quickly.  Bonobo does not respond quickly and allows it to succeed even though it should be failing.

**Comment by AJH16** [comment]
Solved the failure to fail issue.  Apparently the old version of bash used in Bonobo requires the return be exactly 1 to be detected as a non-zero return.  Odd, but workable.  Still getting slow performance though.  This also seems to be an odd edge case as it also revealed a problem in TortoiseGit where the output ends up blank for each line of output even though it works in command line git.  (A blank line shows up for each line of output, but nothing is displayed in TortiseGit.)

**Comment by AJH16** [comment]
Adding some more detail.  It appears that the git process itself is taking a long time to end.  I swapped the order of the copy and the WaitForExit and it takes a long time for WaitForExit to occur.  Perhaps it's something with the particular version of git being used that causes the hook to be called much more slowly or something that prevents it from detecting the exit from the hook.  Any thoughts?

**Comment by willdean** [comment]
I think Bonobo does use a very old version of Git (can't remember off-hand what it is) - I suspect it would be relatively straightforward to upgrade that as an experiment.

**Comment by AJH16** [comment]
It's running 1.9.5 which is the newest version before the 2 cut over which involved a completely different environment type on Windows unfortunately, which is probably why it hasn't been updated yet.  It seems like it would take a pretty substantial amount of work to update it from what I can see.  Trying to swap up the environment was one of the first things I tried, but unfortunately it failed pretty hard (though I didn't yet have quite as good of an understanding of the underlying flow of Bonobo as I've been getting while working through this implementation.)  I know I ran in to issues with newline handling and the way that return codes are handled due to the older bash environments that are used.

**Comment by AJH16** [comment]
Ok, well this is really odd.  Used Procexp to see the process tree and discovered something very unexpected... the issue is not that bash remains open after the process finishes, but rather that bash just sits there without starting execution for the entirety of the delay and then runs the hooks and finishes very quickly.  I'd been focused on the stdout stream, but it looks like it might be something about the stdin stream that is being problematic or else it's some severe issue with the version of bash itself.

**Comment by AJH16** [comment]
Ok, think I have it now... it appears to be an underlying mintty bash shell issue, my best guess is that it is related to the ATI graphics driver issue mentioned here https://stackoverflow.com/questions/42888024/git-bash-mintty-is-extremely-slow-on-windows-10-os.  I was able to replace the bash.exe used by the server and the performance immediately improved (although it resulted in some other oddities now that the newline stuff changed around again).  Our main git server might not have an ATI graphics card though, so I might not actually have to worry about this one.  Will report back when I see if a non-ATI graphics card fixes the issue...

**Comment by willdean** [comment]
Awesome.  The ability of Unix-ported stuff to find weird corner-case behaviour in Windows never fails to impress!

**Comment by AJH16** [comment]
Yeah, confirming that it works fine on the actual server we are using to run Bonobo (a VM with no ATI driver)... go bizarre corner cases!!!
